File: LekkerAndroid/app/src/main/java/com/jacquotq/lekker/service/AIService.kt
package com.jacquotq.lekker.service

import com.google.gson.Gson
import com.google.gson.JsonParser
import com.jacquotq.lekker.data.UserProfile
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException

class AIService {

    // Backend URL — update after deploying Cloudflare Worker
    private val backendUrl = "https://lekker-api.jonstark186.workers.dev/api/query"

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .readTimeout(120, java.util.concurrent.TimeUnit.SECONDS)
        .build()

    private val gson = Gson()

    fun streamQuery(
        query: String,
        profile: UserProfile,
        language: String,
        deviceId: String,
        onToken: (String) -> Unit,
        onDone: () -> Unit,
        onError: (String) -> Unit
    ) {
        val bodyMap = mapOf(
            "query" to query,
            "deviceId" to deviceId,
            "language" to language,
            "profile" to mapOf(
                "etymologyWeight" to profile.etymologyWeight,
                "storiesWeight" to profile.storiesWeight,
                "connectionsWeight" to profile.connectionsWeight,
                "formalWeight" to profile.formalWeight
            )
        )
        val bodyJson = gson.toJson(bodyMap)

        val request = Request.Builder()
            .url(backendUrl)
            .addHeader("content-type", "application/json")
            .post(bodyJson.toRequestBody("application/json".toMediaType()))
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                onError(e.message ?: "Network error")
            }

            override fun onResponse(call: Call, response: Response) {
                if (!response.isSuccessful) {
                    val errBody = response.body?.string() ?: ""
                    if (response.code == 429) {
                        onError("Daily query limit reached. Try again tomorrow.")
                    } else {
                        onError("HTTP ${response.code}: $errBody")
                    }
                    return
                }
                // Parse OpenAI-compatible SSE stream
                response.body?.source()?.use { source ->
                    while (!source.exhausted()) {
                        val line = source.readUtf8Line() ?: break
                        if (!line.startsWith("data: ")) continue
                        val data = line.removePrefix("data: ").trim()
                        if (data == "[DONE]") { onDone(); return }
                        try {
                            val json = JsonParser.parseString(data).asJsonObject
                            val token = json.getAsJsonArray("choices")
                                ?.get(0)?.asJsonObject
                                ?.getAsJsonObject("delta")
                                ?.get("content")?.asString
                            token?.let { onToken(it) }
                        } catch (_: Exception) {}
                    }
                    onDone()
                }
            }
        })
    }
}

File: DutchLearner_iOS/DutchLearner/AIService.swift
import Foundation

// MARK: - AI Service (Backend Proxy)
final class AIService {

    // Backend URL — update after deploying Cloudflare Worker
    private static let backendURL = URL(string: "https://lekker-api.jonstark186.workers.dev/api/query")!

    // ──────────────────────────────────────────────
    // MARK: Streaming Entry Point
    // ──────────────────────────────────────────────
    static func streamChat(
        query: String,
        profile: UserProfile,
        language: String,
        deviceId: String
    ) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    var request = URLRequest(url: backendURL)
                    request.httpMethod = "POST"
                    request.setValue("application/json", forHTTPHeaderField: "Content-Type")

                    let body: [String: Any] = [
                        "query": query,
                        "deviceId": deviceId,
                        "language": language,
                        "profile": [
                            "etymologyWeight": profile.etymologyWeight,
                            "storiesWeight": profile.storiesWeight,
                            "connectionsWeight": profile.connectionsWeight,
                            "formalWeight": profile.formalWeight,
                        ],
                    ]
                    request.httpBody = try JSONSerialization.data(withJSONObject: body)

                    let (bytes, response) = try await URLSession.shared.bytes(for: request)
                    guard let http = response as? HTTPURLResponse else {
                        throw AIServiceError.invalidResponse
                    }
                    if http.statusCode == 429 {
                        throw AIServiceError.rateLimited
                    }
                    guard http.statusCode == 200 else {
                        throw AIServiceError.httpError(http.statusCode)
                    }

                    // Parse OpenAI-compatible SSE stream
                    for try await line in bytes.lines {
                        guard line.hasPrefix("data: ") else { continue }
                        let jsonStr = String(line.dropFirst(6))
                        guard jsonStr != "[DONE]",
                              let data = jsonStr.data(using: .utf8),
                              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                              let choices = obj["choices"] as? [[String: Any]],
                              let first = choices.first,
                              let delta = first["delta"] as? [String: Any],
                              let text = delta["content"] as? String
                        else { continue }
                        continuation.yield(text)
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }
}

// MARK: - Errors
enum AIServiceError: LocalizedError {
    case invalidResponse
    case httpError(Int)
    case rateLimited

    var errorDescription: String? {
        switch self {
        case .invalidResponse:  return String(localized: "error.invalidResponse")
        case .httpError(let c): return String(localized: "error.httpError \(c)")
        case .rateLimited:      return String(localized: "error.rateLimited")
        }
    }
}
